core/packages/contracts/contracts/NativeTokens.sol

```diff
@@ -0,0 +1,123 @@
+    /// @param origin The address which initiated the lock.
+    /// @param recipient The substrate address that will receive the funds.
+    /// @param token The token locked.
+    /// @param amount The amount locked.
+    event Locked(address origin, bytes32 recipient, address token, uint256 amount);
+
+    /// @notice Funds where unlocked.
+    /// @dev Emitted once the funds are unlocked.
+    /// @param origin The substrate address which initiated the unlock. / @param recipient The ethereyn address that will receive the funds.
+    /// @param token The token unlocked.
+    /// @param amount The amount unlocked.
+    event Unlocked(bytes32 origin, address recipient, address token, uint256 amount);
+
+    /// @notice a token was created in Statemint.
+    /// @dev Emitted after enqueueing a a create token message to substrate.
+    event Created(address token);
+
+    /// @dev The vault where ERC20 tokens are locked.
+    ERC20Vault public immutable vault;
+
+    /// @dev The channel used to enqueue messages for lock and create functions.
+    OutboundChannel public immutable outboundChannel;
+
+    /// @notice Initializes the NativeTokens contract with a vault and channels.
+    /// @param _vault The vault to use to `lock`/`unlock` tokens.
+    /// @param _outboundChannel The channel used to queue lock and create messages.
+    constructor(ERC20Vault _vault, OutboundChannel _outboundChannel) {
+        vault = _vault;
+        outboundChannel = _outboundChannel;
+    }
+
+    /// @notice Locks tokens to mint on substrate.
+    /// @dev Locks an amount of ERC20 Tokens in the vault and enqueues a mint message. Requires the allowance to be set on the ERC20 token where the spender is the vault.
+    /// @param token The token to lock.
+    /// @param recipient The recipient on the substrate side.
+    /// @param amount The amount to lock.
+    function lock(address token, bytes32 recipient, uint256 amount) public {
+        require(amount > 0, "NativeTokes: non zero amount");
+        require(token != address(0), "NativeTokes: non zero address");
+        require(recipient != 0, "NativeTokes: non zero recipient");
+
+        /// TODO: Implement a max locked amount.
+        vault.deposit(msg.sender, token, amount);
+
+        /// TODO: Encode a call
+        bytes memory call;
+        /// TODO: Get weight
+        uint64 weight = 1_000_000;
+
+        outboundChannel.submit(msg.sender, call, weight);
+        emit Locked(msg.sender, recipient, token, amount);
+    }
+
+    /// @notice Creates a native token.
+    /// @dev Enqueues a create native token message to substrate.
+    /// @param token The ERC20 token address.
+    /// @param name The name of the ERC20 token.
+    /// @param symbol The symbol of the ERC20 token.
+    /// @param decimals The decimals for the ERC20 token.
+    function create(
+        address token,
+        string calldata name,
+        string calldata symbol,
+        uint8 decimals
```

Comment:
Actually we need to read this metadata from the ERC20 contract, as we can't trust input supplied by an arbitrary user.

https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20Metadata